{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4a70d2b3-ad95-472e-a39c-2c7996357155",
   "metadata": {},
   "source": [
    "## Capstone 2. Predicting Circular Dichorism Spectrum Features from Amino Acid Sequence\n",
    "### Feature Engineering\n",
    "Data is coming from the Protein Circular Dichroism Data Bank (https://pcddb.cryst.bbk.ac.uk/). Citation information: The PCDDB (protein circular dichroism data bank): A bioinformatics resource for protein characterisations and methods development. Ramalli SG, Miles AJ, Janes RW, Wallace BA., J Mol Biol (2022)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "eec0e0b0-4121-4741-8208-6292fd503f58",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import packages\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from aaindex import aaindex1\n",
    "from sklearn.model_selection import train_test_split\n",
    "from scipy.signal import find_peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0aa7ee83-2582-4468-9715-ef8e9a6256fd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Info</th>\n",
       "      <th>CD</th>\n",
       "      <th>Calibration</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CD0000001000</th>\n",
       "      <td>{'PCDDBID': 'CD0000001000', 'FILE EXTRACTED': ...</td>\n",
       "      <td>Wavelength     Final        HT   Smoothed...</td>\n",
       "      <td>Wavelength  Calibration Spectrum\n",
       "0        ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD0000002000</th>\n",
       "      <td>{'PCDDBID': 'CD0000002000', 'FILE EXTRACTED': ...</td>\n",
       "      <td>Wavelength     Final       HT  Smoothed  ...</td>\n",
       "      <td>Wavelength  Calibration Spectrum\n",
       "0        ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD0000003000</th>\n",
       "      <td>{'PCDDBID': 'CD0000003000', 'FILE EXTRACTED': ...</td>\n",
       "      <td>Wavelength     Final     HT  Smoothed  Av...</td>\n",
       "      <td>Wavelength  Calibration Spectrum\n",
       "0        ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD0000004000</th>\n",
       "      <td>{'PCDDBID': 'CD0000004000', 'FILE EXTRACTED': ...</td>\n",
       "      <td>Wavelength     Final     HT  Smoothed  Av...</td>\n",
       "      <td>Wavelength  Calibration Spectrum\n",
       "0        ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD0000005000</th>\n",
       "      <td>{'PCDDBID': 'CD0000005000', 'FILE EXTRACTED': ...</td>\n",
       "      <td>Wavelength     Final        HT   Smoothed...</td>\n",
       "      <td>Wavelength  Calibration Spectrum\n",
       "0        ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                           Info  ...                                        Calibration\n",
       "CD0000001000  {'PCDDBID': 'CD0000001000', 'FILE EXTRACTED': ...  ...      Wavelength  Calibration Spectrum\n",
       "0        ...\n",
       "CD0000002000  {'PCDDBID': 'CD0000002000', 'FILE EXTRACTED': ...  ...      Wavelength  Calibration Spectrum\n",
       "0        ...\n",
       "CD0000003000  {'PCDDBID': 'CD0000003000', 'FILE EXTRACTED': ...  ...      Wavelength  Calibration Spectrum\n",
       "0        ...\n",
       "CD0000004000  {'PCDDBID': 'CD0000004000', 'FILE EXTRACTED': ...  ...      Wavelength  Calibration Spectrum\n",
       "0        ...\n",
       "CD0000005000  {'PCDDBID': 'CD0000005000', 'FILE EXTRACTED': ...  ...      Wavelength  Calibration Spectrum\n",
       "0        ...\n",
       "\n",
       "[5 rows x 3 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Unpickle clean(ish) data\n",
    "dfOG = pd.read_pickle(\"C:/Users/dkoul/OneDrive/Documents/Springboard/Capstone 2/pcd_df.pkl\")\n",
    "dfOG.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d128b530-d4ba-41a0-be51-35e518fc34f4",
   "metadata": {},
   "source": [
    "#### Simplify the above dataframe to only the relevant fields\n",
    "\n",
    "- the PCDDB ids (as index)\n",
    "- amino acid sequence\n",
    "- wavelengths *(select only 190 - 250, then we can disregard this in model and avoid complicating algorithms with experimental settings/buffer interactions/etc.?)*\n",
    "- processed/final cd spectrum *(all converted to $\\Delta$ $\\epsilon$)*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "1d353bcc-ba49-47cc-98e0-56e43484e53b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PCDDB ID</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>CD</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CD0000001000</td>\n",
       "      <td>PHSHPALTPE QKKELSDIAH RIVAPGKGIL AADESTGSIA KR...</td>\n",
       "      <td>0     -0.095282\n",
       "1     -0.150240\n",
       "2     -0.19838...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CD0000002000</td>\n",
       "      <td>RTPEMPVLEN RAAQGDITAP GGARRLTGDQ TAALRDSLSD KP...</td>\n",
       "      <td>0    -0.319511\n",
       "1    -0.332366\n",
       "2    -0.282746\n",
       "3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CD0000003000</td>\n",
       "      <td>ANLNGTLMQY FEWYMPNDGQ HWKRLQNDSA YLAEHGITAV WI...</td>\n",
       "      <td>0     -0.007964\n",
       "1     -0.073311\n",
       "2     -0.10272...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CD0000004000</td>\n",
       "      <td>IVCHTTATSP ISAVTCPPGE NLCYRKMWCD VFCSSRGKVV EL...</td>\n",
       "      <td>0     0.106316\n",
       "1     0.125327\n",
       "2     0.114049\n",
       "3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CD0000005000</td>\n",
       "      <td>CGVPAIQPVL SGLIVNGEEA VPGSWPWQVS LQDKTGFHFC GG...</td>\n",
       "      <td>0    -0.023542\n",
       "1    -0.033641\n",
       "2    -0.039037\n",
       "3...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       PCDDB ID  ...                                                 CD\n",
       "0  CD0000001000  ...  0     -0.095282\n",
       "1     -0.150240\n",
       "2     -0.19838...\n",
       "1  CD0000002000  ...  0    -0.319511\n",
       "1    -0.332366\n",
       "2    -0.282746\n",
       "3...\n",
       "2  CD0000003000  ...  0     -0.007964\n",
       "1     -0.073311\n",
       "2     -0.10272...\n",
       "3  CD0000004000  ...  0     0.106316\n",
       "1     0.125327\n",
       "2     0.114049\n",
       "3...\n",
       "4  CD0000005000  ...  0    -0.023542\n",
       "1    -0.033641\n",
       "2    -0.039037\n",
       "3...\n",
       "\n",
       "[5 rows x 3 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "need_tlc = [72, 74, 98, 121, 482, 508] # indices of values where the metadata fields are jumbled up, identified during EDA\n",
    "\n",
    "# PCDDB ID\n",
    "ids_all = dfOG.index\n",
    "tlc_ids = [ids_all[i] for i in need_tlc]\n",
    "ids = [i for i in ids_all if i not in tlc_ids]\n",
    "\n",
    "# Breakdown of how to access all the information\n",
    "\n",
    "wl = range(190, 251)\n",
    "\n",
    "def extract_data(id):\n",
    "    units = dfOG.loc[id, 'Info']['Dichroism Units of Processed Data']\n",
    "    CD = dfOG.loc[id, 'CD']\n",
    "    if units == 'Delta Epsilon':\n",
    "        spectrum = CD[CD['Wavelength'].isin(wl)]['Final']\n",
    "    elif units == 'Mean Residue Ellipticity':\n",
    "        spectrum = CD[CD['Wavelength'].isin(wl)]['Final']/3298\n",
    "        \n",
    "    seq = dfOG.loc[id, 'Info']['Sequence']\n",
    "    \n",
    "    return spectrum, seq\n",
    "\n",
    "# lists to populate\n",
    "seqs = [] # for amino acid sequences\n",
    "delta_epsilon = [] # spectra for wavelengths 190 - 250\n",
    "errors = [] # for anything that didn't work that I need to look at individually\n",
    "\n",
    "for id in ids:\n",
    "    try:           \n",
    "        spectrum, seq = extract_data(id)\n",
    "        delta_epsilon.append(spectrum.reset_index(drop = True))       \n",
    "        seqs.append(seq)\n",
    "    except:\n",
    "        errors.append(id)\n",
    "\n",
    "# Combine lists into dataframe\n",
    "ids_tlc = tlc_ids + errors\n",
    "ids_notlc = [i for i in ids if i not in ids_tlc]\n",
    "\n",
    "df_clean1 = pd.DataFrame({'PCDDB ID': ids_notlc, 'Sequence': seqs, 'CD': delta_epsilon})\n",
    "\n",
    "# Remove any rows that don't have values for the full wavelength range we're looking at\n",
    "df_clean1['Spectrum Points'] = [len(x) for x in df_clean1['CD']]\n",
    "df_clean2 = df_clean1[df_clean1['Spectrum Points'] == 61]\n",
    "df_clean = df_clean2.drop('Spectrum Points', axis = 1).reset_index(drop = True)\n",
    "\n",
    "df_clean.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbf5518e-58bb-4abc-b2eb-98c3dee50e1f",
   "metadata": {},
   "source": [
    "#### Explanatory variable processing using AAIndex\n",
    "\n",
    "The primary amino acid sequences cannot be used in machine learning algorithms as they are. So, we will use AAIndex to convert the amino acid sequences to numeric vectors based on some physiochemical property, which can then be reduced to a single number as either a mean or sum.\n",
    "\n",
    "https://www.genome.jp/aaindex/\n",
    "Access via DBGET, explained here: https://www.genome.jp/en/about_dbget.html\n",
    "Accessed here via Python package: https://pypi.org/project/aaindex/\n",
    "\n",
    "AAIndex has a LOT of potential amino acid indices. The subset that will be explored here comprises the size, partition coefficient, hydrophobicity, flexibility, chemical shift, helical propensity, and $/phi$ angle. These were chosen as a diverse subset of properties that would be expected to impact secondary structure. The references to the specific indices chosen are specified below in the comment next to each property."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "5f3a6277-66f3-40ad-bb0f-065bf53b9944",
   "metadata": {},
   "outputs": [],
   "source": [
    "size = aaindex1['BIGC670101'].values # Residue volume (Bigelow, 1967)\n",
    "partition_coeff = aaindex1['GARJ730101'].values # Partition coefficient (Garel et al., 1973)\n",
    "hydrophobicity = aaindex1['ARGP820101'].values # Hydrophobicity index (Argos et al., 1982)\n",
    "flexibility = aaindex1['BHAR880101'].values # Average flexibility indices (Bhaskaran-Ponnuswamy, 1988)\n",
    "chemical_shifts = aaindex1['ANDN920101'].values # alpha-CH chemical shifts (Andersen et al., 1992)\n",
    "K_helixcoil = aaindex1['FINA770101'].values #Helix-coil equilibrium constant (Finkelstein-Ptitsyn, 1977)\n",
    "phi = aaindex1['LEVM760104'].values # Side chain torsion angle phi(AAAR) (Levitt, 1976)\n",
    "\n",
    "properties = {'size': size, 'partition_coeff': partition_coeff, 'hydrophobicity': hydrophobicity, \\\n",
    "              'flexibility': flexibility, 'chemical_shifts': chemical_shifts, 'K_helixcoil': K_helixcoil, 'phi': phi}\n",
    "\n",
    "def seq_vectorization(property, df, i):\n",
    "    '''Convert amino acid sequence to numerical vector using AAIndex for property specified.'''\n",
    "    seq = df['Sequence'][i]\n",
    "    num_seq = pd.Series(map(property.get, list(seq)))\n",
    "    return(num_seq.dropna())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "7a173e06-9d5e-463f-8faf-464589f24664",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Iterate over dataframe rows to generate numerical vectors for each amino sequence for all off the chosen properties\n",
    "vect_dict = {'size': [], 'partition_coeff': [], 'hydrophobicity': [], 'flexibility': [], \\\n",
    "             'chemical_shifts': [], 'K_helixcoil': [], 'phi': []}\n",
    "\n",
    "for i in range(len(df_clean)):\n",
    "   for prop in properties:\n",
    "        vect_dict[prop].append(seq_vectorization(properties[prop], df_clean, i))\n",
    "\n",
    "df_propvect = df_clean.join(pd.DataFrame(vect_dict))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "b104caa8-2fab-4cea-ba17-1a799df97998",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Determine mean and sum for each numeric vector representing a sequence\n",
    "\n",
    "df_propvectcols = df_propvect[properties.keys()]\n",
    "\n",
    "means = {}\n",
    "\n",
    "for p in properties.keys():\n",
    "    means[p] = np.zeros(len(df_propvectcols))\n",
    "    for idx, data in enumerate(df_propvectcols[p]):\n",
    "        means[p][idx] = np.mean(data)\n",
    "\n",
    "sums = {}\n",
    "\n",
    "for p in properties.keys():\n",
    "    sums[p] = np.zeros(len(df_propvectcols))\n",
    "    for idx, data in enumerate(df_propvectcols[p]):\n",
    "        sums[p][idx] = np.sum(data)\n",
    "\n",
    "aa_means = pd.DataFrame(means)\n",
    "aa_means['PCDDB ID'] = df_propvect['PCDDB ID']\n",
    "aa_sums = pd.DataFrame(sums)\n",
    "aa_sums['PCDDB ID'] = df_propvect['PCDDB ID']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c16d270-866a-458f-be39-2ecdb273e659",
   "metadata": {},
   "source": [
    "#### Target variable processing: extracting general features from CD spectra\n",
    "\n",
    "Reference: Greenfield NJ \"Using circular dichrosim spectra to estimate protein secondary structure\" Nature Protocols 2006 (doi:10.1038/nprot.2006.202) <p>\n",
    "\n",
    "- $\\alpha$-helices are expressed in negative bands at 222 nm and 208 nm and a positive band at 193 nm\n",
    "- $\\beta$-sheets (antiparallel) are expressed as negative bands 218 nm and positive band at 195 nm\n",
    "- disordered proteins have very low ellipticity above 210 nm and negative bands near 195 nm\n",
    "<p>\n",
    "So we will process the CD data with the following filters to generate three binary target features:<br>\n",
    "As a first pass, we will say that the criteria will be:\n",
    "- minimum between 217 - 219 => $\\beta$\n",
    "- minimum between 220 - 224 => $\\alpha$\n",
    "- negative value at 195 => random coil"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "8b52c6e7-7be8-4251-81d9-149a1ce2529d",
   "metadata": {},
   "outputs": [],
   "source": [
    "''' Correcting erroneous entries where some of the CD entries have wavelengths as part of the CD fields \n",
    "and are formatted as strings. '''\n",
    "\n",
    "CD_corr = np.zeros((len(df_propvect['CD']), 61))\n",
    "\n",
    "for idx, val in enumerate(df_propvect['CD'].values):\n",
    "    try:\n",
    "        CD_corr[idx] = np.array(val).astype('float32')\n",
    "    except ValueError:\n",
    "        val = [s.split()[0] for s in val]\n",
    "        CD_corr[idx] = np.array(val).astype('float32')\n",
    "        \n",
    "target_df_num = pd.DataFrame(CD_corr, columns = wl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "66e5bfbb-5817-4f9f-a1ce-f1edd94e9e1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find wavelengths at which spectrums has minima\n",
    "wl_list = list(target_df_num.columns)\n",
    "minima_at = []\n",
    "for idx, data in target_df_num.iterrows():\n",
    "    min_i, prop = find_peaks(-data)\n",
    "    min_wl = [wl_list[i] for i in min_i]\n",
    "    minima_at.append(min_wl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "fc78d2b5-92dd-462e-a012-6d397e01bdcd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>alpha</th>\n",
       "      <th>beta</th>\n",
       "      <th>random_coil</th>\n",
       "      <th>PCDDB ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>CD0000001000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>CD0000002000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>CD0000003000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>CD0000004000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>CD0000005000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   alpha   beta  random_coil      PCDDB ID\n",
       "0  False   True         True  CD0000001000\n",
       "1   True  False         True  CD0000002000\n",
       "2   True  False         True  CD0000003000\n",
       "3  False  False        False  CD0000004000\n",
       "4  False  False         True  CD0000005000"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Evaluate criteria for alpha helix & beta sheet\n",
    "alpha = np.zeros(len(target_df_num))\n",
    "beta = np.zeros(len(target_df_num))\n",
    "\n",
    "for idx, x in enumerate(minima_at):\n",
    "    for y in x:\n",
    "        if 217 <= y <=219:\n",
    "            beta[idx] = True\n",
    "        if 220 <= y <= 224:\n",
    "            alpha[idx] = True\n",
    "\n",
    "# Evaluate criteria for random coil\n",
    "random_coil = [(x < 0) for x in target_df_num[195]]\n",
    "\n",
    "# Combine into dataframe\n",
    "target_df_cat = pd.DataFrame({'alpha': [bool(x) for x in alpha], 'beta': [bool(x) for x in beta], 'random_coil': random_coil})\n",
    "\n",
    "# add PCDDB ID for reference\n",
    "target_df_cat['PCDDB ID'] = df_propvect['PCDDB ID']\n",
    "\n",
    "target_df_cat.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b76d3f7-36ec-4267-aed4-38e57efcf219",
   "metadata": {},
   "source": [
    "### Conclusions\n",
    "We now have our amino acid sequences encoded as either means ([aa_means]) or sums ([aa_sums]) of their relative index based on six physiochemical properties, hopefully capturing some emergent property or properties of the sequence in one number. Our target variables are three binary categorical features summarizing features of the CD spectra most indicative of secondary structure components (in [target_df_cat]). \n",
    "<p>\n",
    "During EDA, three different modeling algorithms showed the most promise:\n",
    "    - RandomForestClassifier\n",
    "    - AdaBoostClassifier\n",
    "    - SVC (Support Vector Classification)n)\n",
    "</\n",
    "These dataframes will serve as inputs and outputs for the algorithms to either determine an approach to predict broad secondary structural elements from amino acid sequences and/or lend insights into which of the physiochemical properties evaluated are the most deterministic of one or all of the secondary structure elements explored.p>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (Spyder)",
   "language": "python3",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
